File: spec/poc/src/producer.ts
// ─── SDF Producer ─────────────────────────────────────────────────────────────
// Copyright (c) 2026 Yunus YILDIZ — SPDX-License-Identifier: BUSL-1.1
// Implements the canonical producer flow from SDF_FORMAT.md Section 5.
// Steps: validate → generate PDF → assemble JSON files → pack ZIP → return buffer

import { randomUUID } from 'crypto';
import { packContainer } from './container.js';
import { generatePDF } from './pdf.js';
import { validateData } from './validator.js';
import {
  SDFArchive,
  SDFError,
  SDFMeta,
  SDFProducerOptions,
  SDF_ERRORS,
  SDF_VERSION,
} from './types.js';

// ─── buildSDF ─────────────────────────────────────────────────────────────────

export async function buildSDF(options: SDFProducerOptions): Promise<Buffer> {
  const { data, schema, issuer, issuerId, documentType, recipient, recipientId, schemaId, tags } = options;

  // Step 2 — Validate data against schema (SDF_FORMAT.md Section 5, Step 2)
  // Partial or invalid SDF files MUST NOT be written to disk or transmitted.
  const validation = validateData(data, schema);
  if (!validation.valid) {
    throw new SDFError(
      SDF_ERRORS.SCHEMA_MISMATCH,
      'Input data failed schema validation. SDF file was not produced.',
      validation.errors,
      'data.json',
    );
  }

  // Step 3 — Generate visual.pdf (SDF_FORMAT.md Section 5, Step 3)
  const meta = buildMeta({ issuer, issuerId, documentType, recipient, recipientId, schemaId, tags });
  const pdfBytes = await generatePDF({ meta, data });

  // Step 4 — Assemble JSON files (SDF_FORMAT.md Section 5, Step 4)
  const dataJson   = JSON.stringify(data,   null, 2);
  const schemaJson = JSON.stringify(schema, null, 2);
  const metaJson   = JSON.stringify(meta,   null, 2);

  // Step 5 — Pack ZIP archive (SDF_FORMAT.md Section 5, Step 5)
  // TODO: container-decision — currently uses ZIP via packContainer()
  const buffer = await packContainer({
    visualPdf:  pdfBytes,
    dataJson,
    schemaJson,
    metaJson,
  });

  return buffer;
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

function buildMeta(options: Omit<SDFProducerOptions, 'data' | 'schema'>): SDFMeta {
  const meta: SDFMeta = {
    sdf_version:       SDF_VERSION,
    document_id:       randomUUID(),
    issuer:            options.issuer,
    created_at:        new Date().toISOString(),
    // Forward-compatibility marker for Phase 4 (SDF_FORMAT.md Section 4.7)
    signature_algorithm: null,
  };

  if (options.issuerId)      meta.issuer_id        = options.issuerId;
  if (options.documentType)  meta.document_type    = options.documentType;
  if (options.recipient)     meta.recipient        = options.recipient;
  if (options.recipientId)   meta.recipient_id     = options.recipientId;
  if (options.schemaId)      meta.schema_id        = options.schemaId;
  if (options.tags?.length)  meta.tags             = options.tags;

  return meta;
}

// ─── Re-export archive type for consumers ─────────────────────────────────────

export type { SDFArchive };

File: packages/sdf-python/sdf/producer.py
# Copyright (c) 2026 Yunus YILDIZ — SPDX-License-Identifier: BUSL-1.1
"""SDF producer — build .sdf from data and schema. SDF_FORMAT.md Section 5."""

import json
import uuid
import zipfile
from datetime import datetime, timezone
from io import BytesIO
from typing import Any, Optional

from . import _constants
from ._pdf import generate_pdf
from .types import SDFMeta
from .validator import validate_schema_or_throw


def build_sdf(
    data: dict[str, Any],
    schema: dict[str, Any],
    issuer: str,
    *,
    issuer_id: Optional[str] = None,
    document_type: Optional[str] = None,
    recipient: Optional[str] = None,
    recipient_id: Optional[str] = None,
    schema_id: Optional[str] = None,
    tags: Optional[list[str]] = None,
) -> bytes:
    """
    Build an SDF archive from data and schema.
    Returns the .sdf file as bytes.
    """
    validate_schema_or_throw(data, schema)

    meta = SDFMeta(
        sdf_version=_constants.SDF_VERSION,
        document_id=str(uuid.uuid4()),
        issuer=issuer,
        created_at=datetime.now(timezone.utc).isoformat(),
        issuer_id=issuer_id,
        document_type=document_type,
        recipient=recipient,
        recipient_id=recipient_id,
        schema_id=schema_id,
        signature_algorithm=None,
        tags=tags,
    )

    pdf_bytes = generate_pdf(meta, data)

    data_json = json.dumps(data, indent=2)
    schema_json = json.dumps(schema, indent=2)
    meta_json = json.dumps(meta.to_dict(), indent=2)

    buffer = BytesIO()
    with zipfile.ZipFile(buffer, "w", zipfile.ZIP_DEFLATED, compresslevel=6) as zf:
        zf.writestr("visual.pdf", pdf_bytes)
        zf.writestr("data.json", data_json)
        zf.writestr("schema.json", schema_json)
        zf.writestr("meta.json", meta_json)

    return buffer.getvalue()
